# Quant

- 如何交易川普失败的影响？
- [金融证券知识图谱](https://github.com/kevin-meng/financial_stock_knowledge_graph)| |
- [小型金融知识图谱构建流程](https://github.com/jm199504/Financial-Knowledge-Graphs)||
- 模拟交易账户（Paper Trading）测试策略(eg.Interactive Brokers 零售账户)
- [quant_strategy_list5800](https://github.com/fmzquant/strategies?tab=readme-ov-file)|5800 md files|


## 建立自己的系统--认知哲学世界观

你的喜好，你的死亡陷阱

1. 独立交易者能否在市场上长期生存并盈利？ 1)机构体量和反身性; 2)对冲基金之间的策略太相似 -- 许多大型量化基金使用类似的交易策略，比如统计套利、市场中性策略等。
   - 个人交易者如何成长？ 增加交易策略, 不断研究新策略，避免资金增长后策略失效。例如：如果你做日内交易（高频策略），可以尝试长线交易（持仓 1-3 个月）, or进入新的市场，比如期货或加密货币。
   - 交易的关键不是找到“完美策略”，而是灵活适应市场变化，持续优化策略！
1. 炒股的三大原理  
   1. 人的精力是有限的; 
   2. 所有的投资理论都有正确的时候; 
   3. 标的走势由市场的几何结构和动量决定: 市场的几何结构包括图形，成交，比价，市场偏好等，是可以观察和量化的，但同样的几何结构却可以有不同的演变，因为市场动量是参与者当下的合力，是不可预测的，但可预期，操作者唯一需要做的就是根据现实去预期和修正。
2. 怎么识别一本量化交易的书是不是好书。
   - 看作者, 比如《Systematic Trading》的作者Robert，他是一个退休的对冲基金经理，所以他的书观点更加专业，而且经过实践的检验。
   - 《Algorithmic Trading》的作者 Chan，他做过机构投资，但是转为自营，他的书更加偏重个人或者散户投资人，但是相比起来，系统性就会差一些。
   - 《Python for Algorithmic Trading》的作者 Yves，他是学院出身，主要工作是研究和写作，他的书籍显然更加适合初学者，而且写的比较具体和细致，适合没基础的人跟着操作，但是从深度和框架角度会较差一些。
   - [ML Algorith Trading2ed](https://github.com/stefan-jansen/machine-learning-for-trading)|ipynb repository. Stefan是实盘操手
1. 交易策略的来源: 
   - 虽然学术研究可以提供有价值的策略，但许多学术策略可能过于复杂、已经失效（竞争导致利润减少）或需要昂贵的数据才能回测。因此，交易论坛和博客中更简单的策略可能同样有效. e.g. [Wealth-Lab](https://www.wealth-lab.com/)
2. 风险管理中的心理挑战
   - 贪婪（Greed）:当交易表现很好时，加大杠杆，希望快速赚更多钱。
   - 恐惧（Despair）:当交易出现亏损时，因恐惧而过早退出，或者反向加大赌注，希望翻本。
   - 亏损厌恶（Loss Aversion）:过早止盈，却死扛亏损单。
   - 正确做法：1) 设定固定止损（比如亏损 5% 就平仓）。2) 设定浮动止盈（根据市场趋势动态调整目标收益）。



## 常见问题

**回测策略前**(先用简单方法快速筛选，排除明显不可行策略)

1. 是否比市场整体表现更好？
    - Benchmark对比: 10万元投资，你的交易策略能比最简单的投资方式（比如买入指数基金）表现更好？
    - 如果你的策略是多头策略（只做买入），那么它至少要跑赢大盘指数（比如SP500）, 否则不如直接买指数基金
    - 如果你的策略是市场中性（market-neutral），比如对冲基金喜欢做的那种——买入一些股票，同时卖空另一些股票，那么你至少要比无风险收益高，比如美国国债的利息收益。

2. 看两指标: 夏普比率（Sharpe Ratio, SR）, 最大回撤（Max Drawdown, MDD）

3. 幸存者偏差（Survivorship Bias）—— sampling bias
    - 假设你在研究过去 10 年里涨幅最好的 100 只股票，发现它们的平均年回报率高达 30%，你兴奋地想用这些数据回测自己的策略。但这 100 只股票都是“活下来的赢家”，很多早就倒闭的公司被自动忽略了，不在数据里了。
   - 结果：你的策略在历史数据里赚钱，但在现实中会亏损。
   - 尤其要小心低估值选股（Value Investing）策略，因为许多“便宜”的股票后来都倒闭了，历史数据不一定可靠。
   - 解决方案：选择包含退市公司的数据库，比如 CRSP 或 Bloomberg。 自己每天存数据，记录所有股票，包括后来破产的。

4. 很多策略在 10 年前很好用，但现在已经过时了，因为：1) 市场竞争加剧，更多人用相同策略，导致利润消失。2) 市场环境变化，例如法规调整、算法交易普及等。
   - 例子：以前有些策略可以利用卖空限制赚钱，但后来卖空规则放宽，这些策略就失效了。

**回测中常见问题**

1. 怎么评估回测结果? - 两指标
   - 夏普比率: 衡量赚钱的稳定性，告诉你单位风险能赚多少钱。
   - 最大回撤: 衡量亏损的最大风险，告诉你历史上最惨的时候亏了多少。

2. 不同数据源常见问题
    - 拆股 & 分红调整 → 否则股价会“假暴跌”。
    - 幸存者偏差 → 退市的股票可能会影响策略真实表现。
    - 高低点是否可靠 → 避免因假突破/假回撤导致错误交易

3. 前视偏差Look-Ahead Bias, 表现在两个方面: 未来函数(eg偷价, 复权)，和信号闪烁。都是用未来的价格来决定之前的买卖
    - “未来数据作弊”, 你的策略用当天最低价来决定买入，但问题是你在当天交易时根本不知道最低价。这相当于用未来数据作弊，让回测收益看起来比实际更好。
      - [未来函数4个坑](https://zhuanlan.zhihu.com/p/489853249)
    - 解决方案：只使用当时可见的数据，比如前一天的收盘价，而不是当天的最低价。

4. 数据挖掘偏差Data-Snooping Bias or 过度拟合-错误
    - 太复杂的策略容易骗自己，简单才是王道？e.g. 进场信号有 10 个条件，出场信号有 10 个参数……你测试了1000 种均线组合，最终发现9 日均线 + 30 日均线最赚钱“完美”。
    - 问题来了：你是不是“过拟合”了过去的数据。未来市场变化，这些优化的参数根本不适用。
    - 解决方案：
      - 用简单策略，避免过度优化。e.g. 1) 滚动回测（Moving Window Backtesting），看看策略是否在不同时间段都能稳定盈利。2) 经济逻辑 VS. 数学优化。
      - 分训练集和测试集，先在 2010-2015 年优化参数，再用 2016-2020 年测试，看看是否依然有效。选择有逻辑支持的参数，而不是纯粹靠调参优化。

5. 交易成本（Trading Costs）
    - 很多交易策略在纸面上看着很赚钱，但一到实盘就亏损
    - 解决方案：在回测中加入真实交易成本, 测试策略是否仍然能盈利
        - commission = 0.0005  # 手续费（佣金）：每次买卖都会有固定费用 0.05%
        - slippage = 0.0002  # 滑点 0.02% 市场价格可能和你下单的价格不同，比如你想 100 元买入，但市场波动，实际成交价变成了 100.2 元
        - transaction_cost = commission + slippage
    - 市场冲击（market impact）：反身性, 如果你一次买入太多, 反而把价格推高
        - eg.一个基于布林带的策略，在理论测试中表现很好，夏普比率高达 3（非常稳健）。但一旦加入交易成本，夏普比率立刻掉到 -3（严重亏损）。

## 常见策略Alpha

### Overview

|策略	|适用市场	|适用周期	|核心逻辑|
|--| -- | -- | -- |
均值回归	|股票、外汇、期货|	日线/周线	| 价格偏离均值后回归
vs.动量交易	|股票、期货	|1个月-6个月(市场环境) |	强者恒强，涨的更涨
市场状态切换|||怎么判断市场是震荡还是趋势？识别何时转换策略
因子模型||| 量化分析股票涨跌收益的驱动因素
配对交易	| 股票、ETF|	日线/周线	|相关资产价格偏离均值后回归
短期反转	| 股票 |	日内/1-5天	|超跌反弹
日内突破	|美股、A股	|1分钟-1小时 |	突破关键价格后顺势交易
均线交叉	|股票、外汇	|日线/周线	| 金叉买入，死叉卖出
波动率突破 |	期货、外汇|	日线/小时线 | 波动率放大后进入趋势
季节性交易|股票、期货|| 市场都存在季节性规律, “春节效应”或者“圣诞行情”

差异在于交易规则、数据获取方式、执行速度等

市场	| 适用交易	| 主要特点
|--|--|--|
股票（美股、A股）|	价差交易、趋势跟踪	|交易时间固定，波动较大
期货（商品、股指）|	套利、趋势交易	|有杠杆，保证金交易
外汇（Forex）	|高频、趋势交易	|24 小时交易，低成本
加密货币（Crypto）|	高频、套利	|24 小时交易，波动大

策略可分为三大类(核心的量化投资思想)：动量交易、配对交易(套利的一种)和技术分析交易。还根据策略的数据类型（如基本面数据、另类数据）、交易频率（如高频、中低频）、基于套利类型（如期现套利、事件驱动套利）、是否结合机器学习等，来进一步细化和理解这些策略

1. 标的走势由市场的向量和动量决定
   - 向量vector
       1. 技术分析（如价格和趋势信号）: 主要基于图表和技术指标(较主观性), 目标是预测价格走势, 风险是对市场趋势判断的错误, 收益不稳定但可能较高
       2. 统计套利（Statistical Arbitrage）: 基于数学和统计模型(较客观性), 目标是利用价格偏差, 风险是自流动性和模型失效, 收益相对稳定且低
   - 动量 - 事件驱动(eg.新闻信息, 基本面如财务指标)
2. 均值回归(价格回到均衡, 适合套利) vs. 动量交易(跟随趋势)
    - 交易策略一般分为两大类
3. 市场状态切换（Regime Switching）
    - 市场有时是震荡市（适合均值回归），有时是趋势市（适合动量交易）。
    - ✅ 如何判断市场是震荡还是趋势？
        - 低波动（市场上下波动小）→ 市场稳定，适合均值回归策略。
        - 高波动（市场突然波动变大）→ 可能是趋势市，适合动量交易。
        - 使用技术指标ADX（平均方向性指数）, ADX > 25 → 市场有趋势，可以做动量交易。 ADX < 20 → 市场震荡，适合均值回归。
        - 机器学习预测市场变化, eg.隐马尔可夫模型（HMM）：用来检测市场状态是“震荡”还是“趋势”。神经网络：分析大量历史数据，预测市场状态变化。
    - 风险：市场状态很难预测，如果误判市场状态，可能会导致策略失效。

4. 因子模型（Factor Models）, 解释股票价格波动的数学方法，帮助交易者找到影响市场的核心因素。
    - ✅ Fama-French 三因子模型, 经典，认为股票的回报主要由以下三点决定：
        - 市场风险因子（Beta）：股票是否容易跟随市场涨跌。
        - 公司规模因子（Size）：小盘股 vs. 大盘股（小盘股通常表现更好）。
        - 价值因子（Value）：价值股 vs. 成长股（低市净率股票通常更稳定）。
    - 📌 应用：
      - 如果最近小盘股表现好 → 多投资小盘股。
      - 如果价值股跑赢成长股 → 多投资价值股。
    - 📌 风险： 因子收益可能会变化，比如 90 年代互联网泡沫时，成长股大幅跑赢价值股。

5. 交易退出策略
    - 固定持仓时间（例如，动量策略通常持有 1-5 天）。
    - 目标价格: 止盈（例如，盈利 10% 就卖出）。止损策略（例如，亏损 5% 就卖出）。
    - 重新计算交易信号（如果交易策略的信号变了，就退出）。
    - 📌 动量 vs. 均值回归的退出方式
      - 动量交易：趋势反转时退出，或者达到止盈目标时退出。
      - 均值回归交易：当价格回到均值附近，或者持有时间过长都可以退出。
    - 📌 风险：太早止盈会错过更大收益。止损太紧会导致经常被市场震出。

6. 季节性交易（Seasonal Trading）
    - 股票“一月效应” → 小盘股在 1 月通常表现较好。
    - 股票“五月卖出”（Sell in May and go away） → 5-10 月股票市场通常表现较差。
    - 股票“圣诞行情” → 11-12 月市场通常上涨。
    - 期货原油 & 汽油交易：夏季价格通常上涨（旅游需求增加）。
    - 期货农产品：收割季（9-11 月）供给增加，价格下降。
    - 📌 风险：季节性效应可能失效，越来越多交易者知道这个规律，套利空间减少。

> 资金管理 = 资金分配(杠杆) + 风控(退出)策略

无法避免回撤，所以资金管理关键在于"让账户稳定增长 + 如何控制亏损" = "合理利用杠杆 + 防止爆仓(风险)"

1. 资金分配
    - 同时交易3个策略，如何合理分配资金？ Kelly 公式可以帮助你找到最优资金分配方式，确保风险最小，收益最大。
    - Kelly 公式计算出的杠杆：分配较多资金给 策略 A，适当投资 策略 B，并做空策略 C（因为它的收益是负的）。这个方法确保你的整体收益率最大化，同时风险可控。

策略|	平均年收益	|标准差	|计算出的最优杠杆
--|--|--|--
策略 A（石油 ETF）	|13.96%	|10.49%	|1.33
策略 B（银行 ETF）	|2.94%	|6.1%	|0.79
策略 C（零售 ETF）	|-0.73%	|6.48%	|-1.49（应该做空）

2. 风控策略
   - 限制最大回撤（Max Drawdown）
     - 计算历史最大亏损，并设定最大可承受回撤（如不超过 20%）。eg. SP500 最大单日跌幅是 20.47%，你就不应该用超过 1 倍杠杆，否则可能承受超出预期的损失。
   - 使用“半-Kelly”策略
     - Kelly 公式通常计算出的杠杆太激进，所以很多专业交易者使用**“半-Kelly”策略**（把计算出的杠杆减半）。
     - Kelly 计算的杠杆 = 2.528  VS. 实际使用杠杆 = 1.264
   - 动态调整仓位 
     - 亏损时减少杠杆，盈利时合理增加. 如果账户亏损了 10%，Kelly 公式会建议你减少仓位，以免损失进一步扩大。
     - 例如：你有 $100,000 资金，按照 Kelly 公式应该投资 $252,800。如果亏损 10%，账户资金变为 $90,000，应该调整仓位为 $90,000 × 2.528 = $227,520。
   - 交易止损: 
     - 止损也不一定避免爆仓，如市场暴跌，止损单可能无法按预期价格成交，最终亏损更大 eg.2008金融危机
     - 止损只适用于趋势策略（股票突破新高买入），但不适用于均值回归策略（套利策略）



### [打板策略](../wk/qc2strategy.ipynb#打板策略)
### [小市值策略](../wk/qc2strategy.ipynb#小市值策略)
### [消息面-AI文本分析](../wk/qc2strategy.ipynb#消息面-AI文本分析)


# 回测 模拟 交易

**已下载**
1. [stock_market](https://github.com/wbbhcb/stock_market), 待测
1. [alphas](https://github.com/popbo/alphas), 待测ipynb, 利用了这个项目[alphalens](https://github.com/quantopian/alphalens) -> [alphalens-reloaded](https://github.com/stefan-jansen/alphalens-reloaded), [see this ipynb](../wk/0alphalens/alphalens_tutorial_on_quantopian.ipynb#)
1. [blade-fury](https://github.com/albert-feng/blade-fury)| 采集股票数据，利用统计和计算来选股, 要把项目的路径加入PYTHONPATH才能脚本正常运行|
1. [easyquant](https://github.com/shidenggui/easyquant)A股量化框架，支持行情获取以及交易
1. [beginner_event_backtest](https://github.com/szy1900/Event_driven_framework_for_backtesting)

**case**
1. [101Alphas回测比较](https://jglazar.github.io/projects/wq_project/), 最后列出喜欢的5个价格回归比较
2. [QuantsPlaybook](https://github.com/hugo2046/QuantsPlaybook?tab=readme-ov-file), 量化研究-券商金工研报复现

## 数据

1. [AKShare](https://github.com/akfamily/akshare) |elegant/simple financial data library, built for human beings开源财经数据接口库 |
1. [FinnewsHunter](https://github.com/DemonDamon/FinnewsHunter)
    - 从新浪财经、每经网、金融界、中国证券网、证券时报网上，爬取上市公司（个股）的历史新闻文本数据进行文本分析、提取特征集，然后利用SVM、随机森林等分类器进行训练，最后对实施抓取的新闻数据进行分类预测
1. [easyquotation](https://github.com/shidenggui/easyquotation?tab=readme-ov-file)|实时获取免费股票行情，支持新浪 / 腾讯(港股) / 集思录|

数据类型|	作用	|来源/获取方式
--|--|--
历史数据|	回测策略，分析长期趋势|	Yahoo Finance, Quandl, Binance. 数据库查询（MySQL, PostgreSQL）
实时行情（L1 数据）|	计算技术指标，生成买卖信号|	IB API, Binance API
订单簿数据（L2 数据）|	高频交易策略使用 |	深度流动性池（Dark Pool）
财务数据 & 新闻数据|	基本面选股	|Quandl, Alpha Vantage. 爬取网页数据（Scrapy, Selenium）

免费数据源-- ❌ 缺点：通常数据质量一般，可能有幸存者偏差、数据缺失等问题。

|数据来源	| 适用市场	| 数据类型	| 优点	|缺点|
|--| -- | -- | -- |--|
Yahoo Finance | 股票、ETF | 日线数据 |	免费，简单易用	| 不能下载高频数据，可能有幸存者偏差
Alpha Vantage	|股票、外汇|加密	日线、分钟级数据	|免费 API 提供多种数据	|免费额度有限，速度较慢
Binance API |	加密货币|	分钟级数据	|免费,全市场数据	| 只能用于加密货币交易
Quandl（部分免费）|	股票、|经济数据	日线数据	|提供部分免费的宏观经济数据	|高级数据需要付费
Oanda	| 外汇 |	分钟级数据	|免费的外汇数据	| 只能用于外汇市场

付费数据源-- ✅ 优点：数据完整、质量高、包含详细的市场信息（如订单簿、L2 数据）。

|数据来源	| 适用市场	| 数据类型	| 优点	|缺点|
|--| -- | -- | -- |--|
Bloomberg |	全市场	|高频、L2 数据$25,000+/年	|全市场数据，质量高	| 太贵了，个人投资者用不起
Reuters Eikon|	股票、期货、外汇|	高频、新闻数据	$20,000+/年	|专业机构使用	|价格昂贵
TickData	|股票、期货	|Tick 级数据	|按年收费，$1000+	|高频数据质量高	价格较贵
Polygon.io|	股票、加密、外汇	|Tick 级|	月费 $49+	数据完整，支持 L2 订单流|	需要付费
Intrinio	|股票	|财报 + 价格数据 订阅制，$99+	|提供财务因子数据|	价格偏贵



## 回测Backtest 模拟(实盘) 交易live trading

2种回测系统 Vector(speed) .VS. Event(real), 各1套
- 向量化回测框架
    - 优点：1简：策略逻辑可抽象成 pandas/numpy 数据处理; 2快：一次性处理整段时间序列，适合策略参数搜索; 3适合研究阶段的 Alpha 策略验证
    - 缺点：1无法模拟实际撮合细节（委托簿、滑点、延迟）; 2不支持异步事件（如宏观信号、异步市场数据）= 未来函数风险大?; 3难对接真实交易系统
    - 适用策略: 趋势跟踪、因子选股
- 事件驱动回测/实盘框架
    - 优点：1贴近实盘：逐tick/逐bar处理，考虑委托/撮合/滑点; 2支持复杂交易逻辑（如组合管理、风险控制、订单管理系统）; 3和实盘无缝对接
    - 缺点：1慢; 2开发难度大：需要维护事件队列、撮合机制、账户模型
    - 适用策略: 高频交易、做市、套利
    - 支持 Bar → Signal → Order → Execution

> Vector Based Frameworks

1. [FinHack](https://github.com/FinHackCN/finhack)Github库 [快速入门](https://github.com/FinHackCN/finhack/wiki/1%E3%80%81%E5%BF%AB%E9%80%9F%E5%85%A5%E9%97%A8) | 一个易于拓展的量化金融框架，它在当前版本中集成了数据采集、因子计算、因子挖掘、因子分析、机器学习、策略编写、量化回测、实盘接入等全流程的量化投研工作
5. [FastQuant](https://github.com/enzoampil/fastquant) Backtrader封装，新手只需3行代码, 适用于股票、外汇等市场，支持多种技术指标（MACD、均线等）。
6. [bt](https://github.com/pmorissette/bt)简单量化策略优化
7. [vectorbt](https://github.com/polakowo/vectorbt) 5.4k stars | numba | - vectorbt takes a novel approach to backtesting: it operates entirely on pandas and NumPy objects, and is accelerated by Numba to analyze any data at speed and scale. This allows for testing of many thousands of strategies in seconds.
8. [pysystemtrade](https://github.com/robcarver17/pysystemtrade) 2.9k stars| Live Trading | - Systematic Trading in python from book by Rob Carver
9. [finmarketpy](https://github.com/cuemacro/finmarketpy) 3.6k stars| library for backtesting trading strategies & analyzing financial markets (formerly pythalesians)

> Event Driven Frameworks

PS:  live trading marked for at least 1 broker support. Otherwise, backtest function only.

1. [Backtrader](https://github.com/mementum/backtrader) 17.8k stars
   - 开源免费，个人。股票、期货、外汇、加密货币。可接入实盘（IB、Binance、Alpaca等）。内置可视化分析（回测结果、资金曲线、交易信号）
2. [Zipline](https://github.com/stefan-jansen/zipline-reloaded)
   - Quantopian 团队开发，内置数据处理、财报整合Quandl API, 历史模拟与实盘交易集成Alpaca API。适合：研究因子选股、日线级别交易的个人交易者。
3. [autoxd](https://github.com/nessessary/autoxd) | 回测框架, 模拟实盘账户交易, 适合编写T+0策略 |
4. [fooltrader](https://github.com/foolcage/fooltrader) -> [zvt](https://github.com/zvtvz/zvt) | modular quant framework, 项目升级到zvt, 各有详尽解释|
5. [vnpy](https://github.com/vnpy/vnpy) | 基于Python, 开源, 量化交易平台开发框架 
   1. [vnpy example](https://www.zhihu.com/question/356159767)
6. [hikyuu](https://hikyuu.org/)| 开源量化交易研究框架，策略分析, 回测|
7. [Funcat](https://github.com/ricequant/funcat) --> [MyTT](https://github.com/mpquant/MyTT)
     - MyTT将通达信,同花顺,文华麦语言等指标公式,最简移植到Python中,核心库单个文件，仅百行代码,十几个核心函数，实现常见技术指标（不依赖talib库）和转换通达信MACD,RSI,BOLL,ATR,KDJ,CCI,PSY等公式,方便股票指标公式,股市期货量化框架分析,自动程序化交易,数字货币
8. [qstock](https://github.com/tkfy920/qstock)|live trading|包括数据获取（data）、plot、选股(stock)和量化回测（backtest）四个模块
9.  [easytrader](https://github.com/shidenggui/easytrader)|live trading|提供同花顺客户端/miniqmt/雪球的股票量化交易，支持跟踪 joinquant /ricequant 模拟交易 和 实盘雪球组合|
10. [ricequant/RQAlpha](https://github.com/ricequant/rqalpha)|live trading| 数据获取、算法交易、回测引擎，实盘模拟，实盘交易到数据分析 |
11. [QUANTAXIS](https://github.com/yutiansut/QUANTAXIS) |live trading| 任务调度/分布式部署, 股票/期货/期权, 数据/回测/模拟/交易/可视化/多账户 纯本地量化]
       - [QAStrategy](https://github.com/yutiansut/QAStrategy)| QA-sub1
       - [QADemoOLD](https://github.com/QUANTAXIS/QADemoOLD)| QA-sub2 
12. [trader](https://github.com/BigBrotherTrade/trader)|live trading| 操盘大哥的交易组件


In [ ]:
# backtesting_compare: Vector.VS. Event
import numpy as np
import pandas as pd

"""
Minimal demo to compare a vectorized backtester and an event‑driven backtester
running the SAME simple‑moving‑average (SMA) crossover strategy.

Run this file (`python backtesting_compare.py`) to print performance metrics
for both engines. The data is synthetic so the script is self‑contained.
"""

###############################################################################
# Utility helpers
###############################################################################

def generate_data(seed: int = 42, periods: int = 250) -> pd.DataFrame:
    """Create a fake price series so the demo is entirely offline‑capable."""
    rng = np.random.default_rng(seed)
    dates = pd.date_range("2024-01-01", periods=periods, freq="D")
    prices = 100 + rng.standard_normal(periods).cumsum()
    return pd.DataFrame({"Close": prices}, index=dates)

def perf_metrics(equity_curve: pd.Series) -> dict:
    """Return Total Return, Sharpe and Max Drawdown for an equity curve."""
    rets = equity_curve.pct_change().dropna()
    total_return = equity_curve.iloc[-1] / equity_curve.iloc[0] - 1
    sharpe = rets.mean() / rets.std() * np.sqrt(252) if rets.std() != 0 else np.nan
    running_max = equity_curve.cummax()
    drawdown = (equity_curve - running_max) / running_max
    max_dd = drawdown.min()
    return {
        "total_return": round(total_return, 4),
        "sharpe": round(sharpe, 2),
        "max_drawdown": round(max_dd, 4),
    }

###############################################################################
# 1) Vectorized backtester (pandas‑style)
###############################################################################

def vectorized_backtest(
    df: pd.DataFrame, fast: int = 10, slow: int = 30, initial_capital: float = 1_000_000
):
    data = df.copy()
    data["fast"] = data["Close"].rolling(fast).mean()
    data["slow"] = data["Close"].rolling(slow).mean()

    # Trading signal: long when fast MA > slow MA; flat otherwise
    data["position"] = (data["fast"] > data["slow"]).astype(int).shift().fillna(0)

    # Strategy returns (no transaction cost, perfect fills)
    data["strategy_ret"] = data["position"] * data["Close"].pct_change().fillna(0)
    data["equity"] = (1 + data["strategy_ret"]).cumprod() * initial_capital

    return perf_metrics(data["equity"]), data

###############################################################################
# 2) Event‑driven backtester (single‑asset, bar‑by‑bar loop)
###############################################################################

def event_backtest(
    df: pd.DataFrame, fast: int = 10, slow: int = 30, initial_capital: float = 1_000_000
):
    price = df["Close"].values
    fast_ma = pd.Series(price).rolling(fast).mean().values
    slow_ma = pd.Series(price).rolling(slow).mean().values

    equity_curve = []
    cash, position = initial_capital, 0.0  # dollars, shares

    for t in range(len(price)):
        # generate signal when both MAs are ready
        if t >= slow:
            # Cross‑up
            if (
                position == 0
                and fast_ma[t - 1] <= slow_ma[t - 1]
                and fast_ma[t] > slow_ma[t]
            ):
                # BUY at open price of bar t (here use close for simplicity)
                position = cash / price[t]
                cash = 0
            # Cross‑down
            elif (
                position > 0
                and fast_ma[t - 1] >= slow_ma[t - 1]
                and fast_ma[t] < slow_ma[t]
            ):
                # SELL everything
                cash = position * price[t]
                position = 0

        # Equity value mark‑to‑market at bar close
        equity = cash + position * price[t]
        equity_curve.append(equity)

    equity_ser = pd.Series(equity_curve, index=df.index)
    return perf_metrics(equity_ser), equity_ser

###############################################################################
# Demo runner: compare both engines on the same data & strategy
###############################################################################

def main():
    df = generate_data()

    vec_metrics, vec_df = vectorized_backtest(df)
    evt_metrics, evt_curve = event_backtest(df)

    print("=== Vectorized Backtest ===")
    for k, v in vec_metrics.items():
        print(f"{k}: {v}")

    print("\n=== Event‑Driven Backtest ===")
    for k, v in evt_metrics.items():
        print(f"{k}: {v}")

    print("\nDelta (Vectorized – Event):")
    for k in vec_metrics:
        print(f"{k}: {vec_metrics[k] - evt_metrics[k]:.4f}")

if __name__ == "__main__":
    main()


=== Vectorized Backtest ===
total_return: -0.1001
sharpe: -1.17
max_drawdown: -0.1425

=== Event‑Driven Backtest ===
total_return: -0.1189
sharpe: -1.69
max_drawdown: -0.1189

Delta (Vectorized – Event):
total_return: 0.0188
sharpe: 0.5200
max_drawdown: -0.0236


**自动交易系统(Automated Trading System, ATS)** 一套可以自动执行交易策略的计算机程序
   - 市场数据模块（Market Data） → 获取行情数据（价格、成交量、订单簿等）
   - 交易策略模块（Trading Algorithm） → 计算买卖信号
   - 订单执行模块（Order Execution） → 发送交易订单（下单、撤单、修改订单）
   - 风险管理模块（Risk Management） → 控制亏损、监控市场（止损、止盈、仓位管理）
   - 日志 & 监控模块（Logging & Monitoring） → 记录交易 & 复盘, 改进策略

1. 选择交易市场 & 券商: 券商必须提供 API --> 获取市场数据, 分析行情
   - 美股：Interactive Brokers（IB）、TD Ameritrade, TradeStation
   - 期货 & 外汇：Oanda、NinjaTrader、TradeStation
   - 加密货币：Binance、FTX、Kraken
2. 运行交易策略, 生成买卖信号, 交易策略的核心组成: 
   - 信号生成（Signal Generation）：例如均线交叉（10 日均线上穿 50 日均线买入）。
   - 仓位管理（Position Sizing）：计算买入多少股（如最多 10% 账户资金）。
   - 风险控制（Risk Management）：设置止损（-5% 卖出），防止亏损过大。

4. 订单管理（Order Management）, 当策略决定买卖后，系统需要发送交易订单
   - 市价单（Market Order）|	立即成交	|适合高流动性资产
   - 限价单（Limit Order）	|只在特定价格成交	|适合低流动性市场
   - 止损单（Stop Loss）|	亏损超过一定数值后自动卖出	| 适合风险控制

1. 实时监控 & 风险管理: 防止异常情况（滑点、订单失败）。监控哪些风险？
   - 网络问题|	交易失败	|解决：备用服务器
   - 极端市场情况|	突破止损|解决：预设“熔断”机制
   - 账户资金不足|	无法下单|	解决：资金管理系统

2. 日志 & 交易记录: 每笔交易必须记录下来，方便复盘和优化策略。
   - 交易日志包括: 1 交易时间; 2 买卖信号; 3 成交价格; 4 盈利 & 亏损


In [ ]:
# 计算均线交叉信号
import pandas as pd
data["SMA10"] = data["Close"].rolling(window=10).mean()
data["SMA50"] = data["Close"].rolling(window=50).mean()
data["Signal"] = (data["SMA10"] > data["SMA50"]).astype(int)
print(data.tail())

# 通过 IB API 发送订单
from ib_insync import *
ib = IB()
ib.connect('127.0.0.1', 7497, clientId=1)
contract = Stock('AAPL', 'SMART', 'USD')
order = MarketOrder('BUY', 10)
ib.placeOrder(contract, order)

# 设置止损
if data["Close"][-1] < stop_loss_price:
    order = MarketOrder('SELL', 10)
    ib.placeOrder(contract, order)

# 记录交易日志
import logging
logging.basicConfig(filename='trading_log.txt', level=logging.INFO)
logging.info(f"买入 AAPL，价格：{buy_price}，时间：{time_now}")

**Garage:效果不佳, 自测失败或技术底层comments**
1. [mayanghua/instock](https://github.com/myhhub/stock?tab=readme-ov-file) 获取股票数据,计算股票指标,筹码分布,识别股票形态,综合选股,选股策略,股票验证回测,股票自动交易, docker支持PC及移动设备
   - 勉强进入 http://localhost:9988/, add `sudo` in advance
1. QuantConnect ❌ 缺点：本地回测体验不佳，主要依赖云端计算。
1. [Sequoia](https://github.com/sngyai/Sequoia) |好学习样本, A股自动选股实现海龟交易法则、缠中说禅牛市买点技术形态,akshare无法下载 |
1. [ta-lib](https://github.com/TA-Lib/ta-lib-python)|其他包的底层, talib技术指标库 |
1. [backtesting](https://github.com/kernc/backtesting.py)||
1. [qstock](https://github.com/tkfy920/qstock) |付费, depends on pyfolio(too old pkg)|
1. [stock-analysis](https://github.com/wkingnet/stock-analysis) |说明详细值得读，win-based,目前选股条件是阳包阴，而且当天要涨停 |
1. [pyfolio](https://github.com/quantopian/pyfolio)| 5 years ago|
1. [Basana](https://github.com/gbeced/basana)|Python async and event driven framework for algorithmic trading, focus on crypto currencies|


## Terms

1. 趋势策略（股票突破新高买入）
    - 均衡打破, 价格涨了就继续涨，跌了就继续跌
        - eg.股票破新高--某股票过去6个月一直涨，你决定买入，因为它可能继续上涨。直到趋势反转（如股价跌破50日均线）卖出
    - 适合: 小盘股、成长股、加密货币（这些资产容易出现大趋势）
    - 风险: 市场趋势突然反转eg.政策变动、财报暴雷，你的交易可能会被“套牢”。市场上竞争者增加, 回归均衡, 趋势结束
2. 均值回归策略（套利策略）
    - 短期内, 涨太多会跌回均衡值；跌太多会涨回去。
        - eg.股票的布林带（Bollinger Bands）, 某股票过去 100 天的平均价格是 $100，但今天跌到了 $90，你认为它太便宜了，未来可能涨回去，于是买入。当它涨回100以上卖出获利
    - 适合: 震荡市（价格在一个区间内上下波动）。 波动大但长期稳定的市场（eg.大盘股，ETF, crypto）
    - 风险: 如果市场趋势发生变化，比如股票的“合理价格”下降了，你可能会误以为它会反弹，结果继续下跌
3. 策略/因子表达式: 即**Alpha**, 不断拷证alpha的稀缺性/背后的逻辑/规模和持久性
    - Alpha is usually expressed as a number that refers to a percentage. For example, a mutual fund with an alpha of 2.0 would mean that fund has outperformed the market average by 2%
    - Alpha的典型来源: 每个回报来源都来自一个风险源，因此任何投资组合都可以被看作是风险的组合
        - 价格/成交量
        - 基本面
        - 宏观数据
        - 文本：如期刊、新闻、会议记录、社交媒体信息
4. 夏普比率 (Sharpe Ratio)
   - 风险度量: 风险通常用收益率的波动性来表示，波动性越大，风险也就越高。标准差是衡量波动性的一种指标。eg.一个投资组合的年化收益率波动为10%，这10%就代表了你在一年内可能遇到的正负波动幅度。
   - 超额收益: 指的是投资收益减去无风险收益（比如国债收益率）的部分。它反映了你因承担风险而获得的额外收益。
   - 单位风险能赚多少钱: 超额收益/风险（标准差），就得到了每承担一单位风险所获得的超额收益。
   - 收益的波动程度（标准差）Sharpe Ratio = 策略的年收益 - 无风险收益
        - 夏普比率 < 1 ：风险太大，波动太猛，不能单独用。
        - 夏普比率 > 2 ：策略比较稳健，基本能每个月赚钱(每承担1单位的风险（标准差），你可以获得2单位的超额收益)
        - 夏普比率 > 3 ：策略非常稳定，可能是日内高频交易策略，每天都能赚钱。
    - 组合A的表现更好, SR更高:
       - 投资组合A：超额收益为6%，标准差为5%，则夏普比率为 6%/5%=1.2。eg.每个月赚 3%，但每个月都稳定盈利，从不大赚大亏
       - 投资组合B：超额收益为8%，标准差为8%，则夏普比率为 8%/8%=1.0。eg.每个月赚 5%，但有时候赚 10%，有时候亏 5%，波动很大
5. 最大回撤（Maximum Drawdown, MDD）可以用最简单的话来理解：它表示在一段时间内，你的投资或账户价值从最高点到最低点的最大下降幅度
   - 最大亏损幅度：比如你账户从最高点 10 万跌到 7 万，就是 30% 回撤。
   - 最长亏损持续时间：比如这次亏损花了 6 个月才恢复，你能熬得住吗？
6. Kelly formula - 资金管理数学方法，以获得最大化复利增长。计算: 1) 最优杠杆; 2) 资金分配
    - 如果一个策略的平均收益率（期望收益）是 𝑚，标准差（波动性）是 s，那么最优杠杆（该策略的资金比例） 𝑓 = 𝑚/𝑠^2
    - SP500 ETF（SPY）：计算得用 2.5 倍杠杆来投资 SPY，以获得最大复利增长
        - 平均年收益：11.23%
        - 年化波动率（标准差）：16.91%
        - 风险利率：4%
        - Kelly： 𝑓 =（11.23%−4%）/16.91*16.91=2.528
7. 股票的布林带（Bollinger Bands）

### 平台和系统

1. 国际量化冠军赛International Quant Championship (IQC) by WorldQuant is one of the largest and most structured quant competitions globally
   - For early-career quant, IQC is the most accessible and respected global quant tournament with university rankings. But if you're looking to build and monetize strategies, QuantConnect and Numerai are where real-world impact starts.
1. [Composer.trade](https://www.composer.trade/), Alpaca.markets和QuantConnect.com之类的在线交易平台online-trading platform
   - Composer has 11000只股票和ETF中选择，为其构建策略, 构建、测试和押注量化交易算法,图形和AI界面(英语输入请求转化为代码)，
   - 可以让用户以简单流程图的形式查看和分享他们的策略 e.g. 如果标准普尔500指数S&P 500 index最近走高，某种算法会选择持有跟踪纳斯达克100指数的三倍杠杆ETF - triple-leveraged exchange-traded fund tracking the Nasdaq-100 index，反之，则持有美国国债Treasury bills
   - 在Interactive Brokers这类平台上编写交易策略。最近他在Discord上面分享策略后得到了一个想法half-baked idea，用跟踪半导体类股的ETF取代了跟踪纳斯达克100指数的ETF，从而改善了其中一项交易策略的表现
1. [Quantconnect](https://github.com/QuantConnect): platform for quant trading
   - `docker pull quantconnect/lean` or `quantconnect/research` 
   - Quantconnect lean: Lean is the open-source algorithmic trading engine that powers QuantConnect
   - Lean CLI: 
     - is a command interface for running lean engine, and to allow syncing btw local and remote projects
     - can be installed and updated by running `pip install --upgrade lean`(Python package)

高频
1. 高频HFT数据管理
   - [ArcticDB: high performance, serverless DataFrame database for Python Data ecosystem](https://github.com/man-group/ArcticDB)
   - [opentsdb: free scalable, distributed Time Series Database](https://github.com/OpenTSDB/opentsdb)
   - [timescaledb: time-series database for high-performance real-time analytics packaged as a Postgres extension](https://github.com/timescale/timescaledb)

**Reference:**
- [awesome-quant](https://github.com/thuquant/awesome-quant)中国的Quant相关资源索引
- [awesome-stars](https://github.com/dick318/awesome-stars)按语言分类项目, Github
- [awesome-systematic-trading](https://github.com/wangzhe3224/awesome-systematic-trading)筛选过，有General Alpha, 分类特别好